### PR TITLE
feat(lua): deprecate vim.tbl_add_reverse_lookup

### PR DESCRIPTION
**Pull request by MariaSolOs**
Inspired by #24564, deprecating and removing our usages of this troublesome function.

### CODE REVIEW

**Review by lewis6991 — APPROVED** [review]

**Comment by zeertzjq** [comment]
This is not a breaking change.

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -459,6 +459,7 @@ end
 --- ``tbl_add_reverse_lookup { A = 1 } == { [1] = 'A', A = 1 }``
 ---
 --- Note that this *modifies* the input.
+---@deprecated
```
Comment (wookayin) [inline runtime/lua/vim/shared.lua:462]: Can we add vim.deprecate() warning?

**Comment by MariaSolOs** [comment]
Ehhh what's going on with the docs check?

**Comment by MariaSolOs** [comment]
I would like someone to explain me why https://github.com/neovim/neovim/pull/27639/commits/cbdb3f5f34d079c0de0057d286fbe2a10fbab97f fixes [these CI failures](https://github.com/neovim/neovim/actions/runs/8057746383/job/22009414828?pr=27639), because I'm confused.

File: runtime/lua/vim/lsp/log.lua [hunk 2]
```diff
@@ -80,9 +81,10 @@ end
 for level, levelnr in pairs(log.levels) do
   -- Also export the log level on the root object.
   log[level] = levelnr
-end
 
-vim.tbl_add_reverse_lookup(log.levels)
+  -- Add a reverse lookup.
```
Comment (clason) [inline runtime/lua/vim/lsp/log.lua:85]: My point was actually to _not_ shoehorn the reverse lookup into the _same_ table. Can we make this a separate table?

Comment (clason) [reply runtime/lua/vim/lsp/log.lua:85]: (And same below)

Comment (MariaSolOs) [reply runtime/lua/vim/lsp/log.lua:87]: @clason please let me know if I've addressed your concerns :)

Comment (clason) [reply runtime/lua/vim/lsp/log.lua:87]: looks good!

**Comment by lewis6991** [comment]
CI looks sad.

**Comment by MariaSolOs** [comment]
> CI looks sad.

Yeah, I'm waiting for it to cheer up.

**Comment by MariaSolOs** [comment]
Habemus green CI!

**Review by lewis6991 — APPROVED** [review]

**Comment by clason** [comment]
And good riddance!

**Comment by wookayin** [comment]
For those who might find this issue by searching the web: to avoid using the deprecated function, one might write a (simple) for loop. Here is an alternative implementation:

```lua
for _, k in ipairs(vim.tbl_keys(t)) do
  t[t[k]] = k
end
```

Note: Do not update the same table while iterating through it. It will often work incorrectly:
```lua
-- XXX this is buggy!
for k, v in pairs(t) do
  t[v] = k
end
```

**Pull request merged**